File: python/src/transliteration_core.py
from __future__ import annotations

import base64
import binascii
from collections.abc import Iterable, Mapping
from dataclasses import dataclass, field
from enum import Enum
import json
from typing import Any
import unicodedata


class UnicodeNormalizationForm(str, Enum):
    PRESERVE = 'preserve'
    NFC = 'nfc'
    NFD = 'nfd'

    # Dart-style aliases
    preserve = PRESERVE
    nfc = NFC
    nfd = NFD


class TransliterationProfile(str, Enum):
    STRICT_IAST = 'strictIast'
    ISO_15919_CORE = 'iso15919Core'
    EXTENDED_INDIC = 'extendedIndic'
    HUNTERIAN = 'hunterian'
    PLAIN_ENGLISH = 'plainEnglish'

    strictIast = STRICT_IAST
    iso15919Core = ISO_15919_CORE
    extendedIndic = EXTENDED_INDIC
    hunterian = HUNTERIAN
    plainEnglish = PLAIN_ENGLISH


class TransliterationIssueSeverity(str, Enum):
    INFO = 'info'
    WARNING = 'warning'
    ERROR = 'error'

    info = INFO
    warning = WARNING
    error = ERROR


@dataclass(frozen=True, slots=True)
class TransliterationIssue:
    code: str
    message: str
    severity: TransliterationIssueSeverity = TransliterationIssueSeverity.WARNING
    source_rune_offset: int | None = None

    @property
    def sourceRuneOffset(self) -> int | None:  # Dart-style alias
        return self.source_rune_offset

    def to_json(self) -> dict[str, Any]:
        return {
            'code': self.code,
            'message': self.message,
            'severity': self.severity.value,
            'sourceRuneOffset': self.source_rune_offset,
        }

    def toJson(self) -> dict[str, Any]:
        return self.to_json()

    @classmethod
    def from_json(cls, value: Mapping[str, Any]) -> TransliterationIssue:
        return cls(
            code=str(value['code']),
            message=str(value['message']),
            severity=TransliterationIssueSeverity(str(value['severity'])),
            source_rune_offset=value.get('sourceRuneOffset'),
        )

    fromJson = from_json


@dataclass(frozen=True, slots=True)
class LosslessTransliterationResult:
    original: str
    normalized_input: str
    rendered: str
    profile: TransliterationProfile
    input_normalization: UnicodeNormalizationForm
    output_normalization: UnicodeNormalizationForm
    rendering_is_injective: bool
    issues: tuple[TransliterationIssue, ...] = field(default_factory=tuple)

    @property
    def normalizedInput(self) -> str:
        return self.normalized_input

    @property
    def inputNormalization(self) -> UnicodeNormalizationForm:
        return self.input_normalization

    @property
    def outputNormalization(self) -> UnicodeNormalizationForm:
        return self.output_normalization

    @property
    def renderingIsInjective(self) -> bool:
        return self.rendering_is_injective

    @property
    def original_code_points(self) -> tuple[int, ...]:
        return tuple(ord(ch) for ch in self.original)

    @property
    def originalCodePoints(self) -> tuple[int, ...]:
        return self.original_code_points

    def restore_original(self) -> str:
        return self.original

    def restoreOriginal(self) -> str:
        return self.restore_original()

    @property
    def exact_source_recovery_available(self) -> bool:
        return True

    @property
    def exactSourceRecoveryAvailable(self) -> bool:
        return True

    @property
    def has_errors(self) -> bool:
        return any(issue.severity is TransliterationIssueSeverity.ERROR for issue in self.issues)

    @property
    def hasErrors(self) -> bool:
        return self.has_errors

    def to_json(self) -> dict[str, Any]:
        return {
            'schema': 'indic-script-converter/1',
            'original': self.original,
            'originalCodePoints': list(self.original_code_points),
            'normalizedInput': self.normalized_input,
            'rendered': self.rendered,
            'profile': self.profile.value,
            'inputNormalization': self.input_normalization.value,
            'outputNormalization': self.output_normalization.value,
            'renderingIsInjective': self.rendering_is_injective,
            'issues': [issue.to_json() for issue in self.issues],
        }

    def toJson(self) -> dict[str, Any]:
        return self.to_json()

    @classmethod
    def from_json(cls, value: Mapping[str, Any]) -> LosslessTransliterationResult:
        if value.get('schema') != 'indic-script-converter/1':
            raise ValueError('Unsupported transliteration envelope.')

        original = str(value['original'])
        encoded = [int(v) for v in value['originalCodePoints']]
        actual = [ord(ch) for ch in original]
        if encoded != actual:
            raise ValueError('Envelope source code-point integrity check failed.')

        return cls(
            original=original,
            normalized_input=str(value['normalizedInput']),
            rendered=str(value['rendered']),
            profile=TransliterationProfile(str(value['profile'])),
            input_normalization=UnicodeNormalizationForm(str(value['inputNormalization'])),
            output_normalization=UnicodeNormalizationForm(str(value['outputNormalization'])),
            rendering_is_injective=bool(value['renderingIsInjective']),
            issues=tuple(TransliterationIssue.from_json(v) for v in value['issues']),
        )

    fromJson = from_json

    def to_json_text(
        self, *, ensure_ascii: bool = False, separators: tuple[str, str] | None = None
    ) -> str:
        kwargs: dict[str, Any] = {'ensure_ascii': ensure_ascii}
        if separators is not None:
            kwargs['separators'] = separators
        return json.dumps(self.to_json(), **kwargs)

    @classmethod
    def from_json_text(cls, text: str) -> LosslessTransliterationResult:
        value = json.loads(text)
        if not isinstance(value, dict):
            raise ValueError('Transliteration envelope must be a JSON object.')
        return cls.from_json(value)


def normalize_unicode(text: str, form: UnicodeNormalizationForm) -> str:
    if form is UnicodeNormalizationForm.PRESERVE:
        return text
    if form is UnicodeNormalizationForm.NFC:
        return unicodedata.normalize('NFC', text)
    if form is UnicodeNormalizationForm.NFD:
        return unicodedata.normalize('NFD', text)
    raise ValueError(f'Unsupported normalization form: {form!r}')


normalizeUnicode = normalize_unicode


def is_unicode_combining_mark(value: int | str) -> bool:
    ch = chr(value) if isinstance(value, int) else value
    if len(ch) != 1:
        return False
    return unicodedata.category(ch) in {'Mn', 'Mc', 'Me'}


isUnicodeCombiningMark = is_unicode_combining_mark


def is_encoded_vedic_mark(value: int | str) -> bool:
    cp = ord(value) if isinstance(value, str) else value
    return cp in {0x0951, 0x0952} or 0x1CD0 <= cp <= 0x1CFF or 0xA8E0 <= cp <= 0xA8FF


isEncodedVedicMark = is_encoded_vedic_mark


@dataclass(frozen=True, slots=True)
class EmbeddedExactSource:
    visible_text: str
    original_source: str

    @property
    def visibleText(self) -> str:
        return self.visible_text

    @property
    def originalSource(self) -> str:
        return self.original_source


_EXACT_SOURCE_START_TAG = 0xE0001
_EXACT_SOURCE_END_TAG = 0xE007F
_EXACT_SOURCE_MAGIC = 'LIT1:'


def _string_to_utf16le(text: str) -> bytes:
    # surrogatepass is required to match Dart's exact UTF-16 code-unit storage.
    return text.encode('utf-16-le', errors='surrogatepass')


def _string_from_utf16le(data: bytes) -> str:
    if len(data) % 2:
        raise ValueError('Invalid UTF-16LE source payload length.')
    return data.decode('utf-16-le', errors='surrogatepass')


def _fnv1a32(data: bytes | bytearray | memoryview | Iterable[int]) -> int:
    h = 0x811C9DC5
    for byte in data:
        h ^= int(byte)
        h = (h * 0x01000193) & 0xFFFFFFFF
    return h


def embed_exact_source_metadata(rendered: str, original_source: str) -> str:
    source_bytes = _string_to_utf16le(original_source)
    encoded = base64.urlsafe_b64encode(source_bytes).decode('ascii').rstrip('=')
    source_checksum = f'{_fnv1a32(source_bytes):08x}'
    rendered_checksum = f'{_fnv1a32(_string_to_utf16le(rendered)):08x}'
    payload = f'{_EXACT_SOURCE_MAGIC}{encoded}:{source_checksum}:{rendered_checksum}'

    tagged = [_EXACT_SOURCE_START_TAG]
    for unit in payload.encode('ascii'):
        if unit < 0x20 or unit > 0x7E:
            raise RuntimeError('Exact-source payload unexpectedly contains non-ASCII.')
        tagged.append(0xE0000 + unit)
    tagged.append(_EXACT_SOURCE_END_TAG)
    return rendered + ''.join(chr(cp) for cp in tagged)


embedExactSourceMetadata = embed_exact_source_metadata


def try_decode_exact_source_metadata(text: str) -> EmbeddedExactSource | None:
    if not text or ord(text[-1]) != _EXACT_SOURCE_END_TAG:
        return None

    start = len(text) - 2
    while start >= 0 and ord(text[start]) != _EXACT_SOURCE_START_TAG:
        start -= 1
    if start < 0:
        return None

    payload_units: list[int] = []
    for ch in text[start + 1 : -1]:
        cp = ord(ch)
        if cp < 0xE0020 or cp > 0xE007E:
            return None
        payload_units.append(cp - 0xE0000)

    try:
        payload = bytes(payload_units).decode('ascii')
    except UnicodeDecodeError:
        return None
    if not payload.startswith(_EXACT_SOURCE_MAGIC):
        return None

    body = payload[len(_EXACT_SOURCE_MAGIC) :]
    rendered_split = body.rfind(':')
    if rendered_split <= 0 or rendered_split == len(body) - 1:
        return None
    source_split = body.rfind(':', 0, rendered_split)
    if source_split <= 0 or source_split == rendered_split - 1:
        return None

    encoded = body[:source_split]
    source_checksum = body[source_split + 1 : rendered_split]
    rendered_checksum = body[rendered_split + 1 :]
    if len(source_checksum) != 8 or len(rendered_checksum) != 8:
        return None
    try:
        int(source_checksum, 16)
        int(rendered_checksum, 16)
    except ValueError:
        return None

    try:
        padded = encoded + '=' * ((4 - len(encoded) % 4) % 4)
        source_bytes = base64.urlsafe_b64decode(padded.encode('ascii'))
        if f'{_fnv1a32(source_bytes):08x}' != source_checksum:
            return None
        source = _string_from_utf16le(source_bytes)
        visible = text[:start]
        if f'{_fnv1a32(_string_to_utf16le(visible)):08x}' != rendered_checksum:
            return None
        return EmbeddedExactSource(visible_text=visible, original_source=source)
    except (ValueError, UnicodeError, binascii.Error):
        return None


tryDecodeExactSourceMetadata = try_decode_exact_source_metadata


def strip_exact_source_metadata(text: str) -> str:
    decoded = try_decode_exact_source_metadata(text)
    return decoded.visible_text if decoded is not None else text


stripExactSourceMetadata = strip_exact_source_metadata


def recover_embedded_exact_source(text: str) -> str | None:
    decoded = try_decode_exact_source_metadata(text)
    return decoded.original_source if decoded is not None else None


recoverEmbeddedExactSource = recover_embedded_exact_source


def has_embedded_exact_source(text: str) -> bool:
    return try_decode_exact_source_metadata(text) is not None


hasEmbeddedExactSource = has_embedded_exact_source


__all__ = [
    'EmbeddedExactSource',
    'LosslessTransliterationResult',
    'TransliterationIssue',
    'TransliterationIssueSeverity',
    'TransliterationProfile',
    'UnicodeNormalizationForm',
    'embedExactSourceMetadata',
    'embed_exact_source_metadata',
    'hasEmbeddedExactSource',
    'has_embedded_exact_source',
    'isEncodedVedicMark',
    'isUnicodeCombiningMark',
    'is_encoded_vedic_mark',
    'is_unicode_combining_mark',
    'normalizeUnicode',
    'normalize_unicode',
    'recoverEmbeddedExactSource',
    'recover_embedded_exact_source',
    'stripExactSourceMetadata',
    'strip_exact_source_metadata',
    'tryDecodeExactSourceMetadata',
    'try_decode_exact_source_metadata',
]

File: javascript/src/transliteration-core.js
/**
 * Core exact round-trip transliteration primitives.
 *
 * This module intentionally mirrors the Dart/Python envelope schema and the
 * invisible Unicode Tag metadata format byte-for-byte.
 */

export const UnicodeNormalizationForm = Object.freeze({
    PRESERVE: 'preserve',
    NFC: 'nfc',
    NFD: 'nfd',
    preserve: 'preserve',
    nfc: 'nfc',
    nfd: 'nfd',
});

export const TransliterationProfile = Object.freeze({
    STRICT_IAST: 'strictIast',
    ISO_15919_CORE: 'iso15919Core',
    EXTENDED_INDIC: 'extendedIndic',
    HUNTERIAN: 'hunterian',
    PLAIN_ENGLISH: 'plainEnglish',
    strictIast: 'strictIast',
    iso15919Core: 'iso15919Core',
    extendedIndic: 'extendedIndic',
    hunterian: 'hunterian',
    plainEnglish: 'plainEnglish',
});

export const TransliterationIssueSeverity = Object.freeze({
    INFO: 'info',
    WARNING: 'warning',
    ERROR: 'error',
    info: 'info',
    warning: 'warning',
    error: 'error',
});

export class TransliterationIssue {
    constructor({
        code,
        message,
        severity = TransliterationIssueSeverity.WARNING,
        sourceRuneOffset = null,
        source_rune_offset = sourceRuneOffset,
    }) {
        this.code = String(code);
        this.message = String(message);
        this.severity = severity;
        this.sourceRuneOffset = source_rune_offset;
        this.source_rune_offset = source_rune_offset;
        Object.freeze(this);
    }

    toJson() {
        return {
            code: this.code,
            message: this.message,
            severity: this.severity,
            sourceRuneOffset: this.sourceRuneOffset,
        };
    }

    to_json() {
        return this.toJson();
    }

    static fromJson(value) {
        return new TransliterationIssue({
            code: value.code,
            message: value.message,
            severity: value.severity,
            sourceRuneOffset: value.sourceRuneOffset ?? null,
        });
    }

    static from_json(value) {
        return TransliterationIssue.fromJson(value);
    }
}

export class TransliterationResult {
    constructor({
        original,
        normalizedInput,
        normalized_input = normalizedInput,
        rendered,
        profile,
        inputNormalization,
        input_normalization = inputNormalization,
        outputNormalization,
        output_normalization = outputNormalization,
        renderingIsInjective,
        rendering_is_injective = renderingIsInjective,
        issues = [],
    }) {
        this.original = String(original);
        this.normalizedInput = String(normalized_input);
        this.normalized_input = this.normalizedInput;
        this.rendered = String(rendered);
        this.profile = profile;
        this.inputNormalization = input_normalization;
        this.input_normalization = this.inputNormalization;
        this.outputNormalization = output_normalization;
        this.output_normalization = this.outputNormalization;
        this.renderingIsInjective = Boolean(rendering_is_injective);
        this.rendering_is_injective = this.renderingIsInjective;
        this.issues = Object.freeze(Array.from(issues));
        Object.freeze(this);
    }

    get originalCodePoints() {
        return Array.from(this.original, (ch) => ch.codePointAt(0));
    }

    get original_code_points() {
        return this.originalCodePoints;
    }

    restoreOriginal() {
        return this.original;
    }

    restore_original() {
        return this.original;
    }

    get exactSourceRecoveryAvailable() {
        return true;
    }

    get exact_source_recovery_available() {
        return true;
    }

    get hasErrors() {
        return this.issues.some((issue) => issue.severity === TransliterationIssueSeverity.ERROR);
    }

    get has_errors() {
        return this.hasErrors;
    }

    toJson() {
        return {
            schema: 'exact round-trip-indic-transliteration/1',
            original: this.original,
            originalCodePoints: this.originalCodePoints,
            normalizedInput: this.normalizedInput,
            rendered: this.rendered,
            profile: this.profile,
            inputNormalization: this.inputNormalization,
            outputNormalization: this.outputNormalization,
            renderingIsInjective: this.renderingIsInjective,
            issues: this.issues.map((issue) =>
                issue instanceof TransliterationIssue ? issue.toJson() : issue,
            ),
        };
    }

    to_json() {
        return this.toJson();
    }

    toJsonText(space = 0) {
        return JSON.stringify(this.toJson(), null, space);
    }

    to_json_text(space = 0) {
        return this.toJsonText(space);
    }

    static fromJson(value) {
        if (value?.schema !== 'exact round-trip-indic-transliteration/1') {
            throw new TypeError('Unsupported transliteration envelope.');
        }

        const original = String(value.original);
        const encodedCodePoints = Array.from(value.originalCodePoints, Number);
        const actualCodePoints = Array.from(original, (ch) => ch.codePointAt(0));
        if (!sameIntegers(encodedCodePoints, actualCodePoints)) {
            throw new TypeError('Envelope source code-point integrity check failed.');
        }

        return new TransliterationResult({
            original,
            normalizedInput: String(value.normalizedInput),
            rendered: String(value.rendered),
            profile: value.profile,
            inputNormalization: value.inputNormalization,
            outputNormalization: value.outputNormalization,
            renderingIsInjective: Boolean(value.renderingIsInjective),
            issues: Array.from(value.issues ?? [], (item) => TransliterationIssue.fromJson(item)),
        });
    }

    static from_json(value) {
        return TransliterationResult.fromJson(value);
    }

    static fromJsonText(text) {
        const value = JSON.parse(text);
        if (value === null || typeof value !== 'object' || Array.isArray(value)) {
            throw new TypeError('Transliteration envelope must be a JSON object.');
        }
        return TransliterationResult.fromJson(value);
    }

    static from_json_text(text) {
        return TransliterationResult.fromJsonText(text);
    }
}

function sameIntegers(a, b) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i += 1) {
        if (a[i] !== b[i]) return false;
    }
    return true;
}

export function normalizeUnicode(input, form) {
    const text = String(input);
    switch (form) {
        case UnicodeNormalizationForm.PRESERVE:
            return text;
        case UnicodeNormalizationForm.NFC:
            return text.normalize('NFC');
        case UnicodeNormalizationForm.NFD:
            return text.normalize('NFD');
        default:
            throw new RangeError(`Unsupported normalization form: ${String(form)}`);
    }
}

export const normalize_unicode = normalizeUnicode;

const COMBINING_MARK_RE = /^\p{M}$/u;

export function isUnicodeCombiningMark(value) {
    let ch;
    if (typeof value === 'number') {
        try {
            ch = String.fromCodePoint(value);
        } catch {
            return false;
        }
    } else {
        ch = String(value);
    }
    return Array.from(ch).length === 1 && COMBINING_MARK_RE.test(ch);
}

export const is_unicode_combining_mark = isUnicodeCombiningMark;

export function isEncodedVedicMark(value) {
    const cp = typeof value === 'number' ? value : String(value).codePointAt(0);
    return (
        cp === 0x0951 ||
        cp === 0x0952 ||
        (cp >= 0x1cd0 && cp <= 0x1cff) ||
        (cp >= 0xa8e0 && cp <= 0xa8ff)
    );
}

export const is_encoded_vedic_mark = isEncodedVedicMark;

export class EmbeddedExactSource {
    constructor({
        visibleText,
        visible_text = visibleText,
        originalSource,
        original_source = originalSource,
    }) {
        this.visibleText = String(visible_text);
        this.visible_text = this.visibleText;
        this.originalSource = String(original_source);
        this.original_source = this.originalSource;
        Object.freeze(this);
    }
}

const EXACT_SOURCE_START_TAG = 0xe0001;
const EXACT_SOURCE_END_TAG = 0xe007f;
const EXACT_SOURCE_MAGIC = 'LIT1:';

function stringToUtf16Le(text) {
    return Buffer.from(String(text), 'utf16le');
}

function stringFromUtf16Le(bytes) {
    const buffer = Buffer.from(bytes);
    if (buffer.length % 2 !== 0) {
        throw new TypeError('Invalid UTF-16LE source payload length.');
    }
    return buffer.toString('utf16le');
}

function fnv1a32(bytes) {
    let hash = 0x811c9dc5;
    for (const byte of bytes) {
        hash ^= byte;
        hash = Math.imul(hash, 0x01000193) >>> 0;
    }
    return hash >>> 0;
}

function checksumHex(bytes) {
    return fnv1a32(bytes).toString(16).padStart(8, '0');
}

function codePointsToString(points) {
    const chunks = [];
    const chunkSize = 8192;
    for (let i = 0; i < points.length; i += chunkSize) {
        chunks.push(String.fromCodePoint(...points.slice(i, i + chunkSize)));
    }
    return chunks.join('');
}

export function embedExactSourceMetadata(rendered, originalSource) {
    const visible = String(rendered);
    const source = String(originalSource);
    const bytes = stringToUtf16Le(source);
    const encoded = bytes.toString('base64url');
    const sourceChecksum = checksumHex(bytes);
    const renderedChecksum = checksumHex(stringToUtf16Le(visible));
    const payload = `${EXACT_SOURCE_MAGIC}${encoded}:${sourceChecksum}:${renderedChecksum}`;

    const taggedPayload = [EXACT_SOURCE_START_TAG];
    for (const unit of Array.from(payload, (ch) => ch.codePointAt(0))) {
        if (unit < 0x20 || unit > 0x7e) {
            throw new Error('Exact-source payload unexpectedly contains non-ASCII.');
        }
        taggedPayload.push(0xe0000 + unit);
    }
    taggedPayload.push(EXACT_SOURCE_END_TAG);
    return visible + codePointsToString(taggedPayload);
}

export const embed_exact_source_metadata = embedExactSourceMetadata;

export function tryDecodeExactSourceMetadata(text) {
    const sourceText = String(text);
    const runes = Array.from(sourceText, (ch) => ch.codePointAt(0));
    if (runes.length === 0 || runes.at(-1) !== EXACT_SOURCE_END_TAG) return null;

    let start = runes.length - 2;
    while (start >= 0 && runes[start] !== EXACT_SOURCE_START_TAG) start -= 1;
    if (start < 0) return null;

    const payloadUnits = [];
    for (let i = start + 1; i < runes.length - 1; i += 1) {
        const rune = runes[i];
        if (rune < 0xe0020 || rune > 0xe007e) return null;
        payloadUnits.push(rune - 0xe0000);
    }

    const payload = codePointsToString(payloadUnits);
    if (!payload.startsWith(EXACT_SOURCE_MAGIC)) return null;

    const body = payload.slice(EXACT_SOURCE_MAGIC.length);
    const renderedSplit = body.lastIndexOf(':');
    if (renderedSplit <= 0 || renderedSplit === body.length - 1) return null;
    const sourceSplit = body.lastIndexOf(':', renderedSplit - 1);
    if (sourceSplit <= 0 || sourceSplit === renderedSplit - 1) return null;

    const encoded = body.slice(0, sourceSplit);
    const sourceChecksumText = body.slice(sourceSplit + 1, renderedSplit);
    const renderedChecksumText = body.slice(renderedSplit + 1);
    if (!/^[0-9a-f]{8}$/.test(sourceChecksumText)) return null;
    if (!/^[0-9a-f]{8}$/.test(renderedChecksumText)) return null;

    try {
        const bytes = Buffer.from(encoded, 'base64url');
        if (checksumHex(bytes) !== sourceChecksumText) return null;
        const originalSource = stringFromUtf16Le(bytes);
        const visibleText = codePointsToString(runes.slice(0, start));
        if (checksumHex(stringToUtf16Le(visibleText)) !== renderedChecksumText) {
            return null;
        }
        return new EmbeddedExactSource({ visibleText, originalSource });
    } catch {
        return null;
    }
}

export const try_decode_exact_source_metadata = tryDecodeExactSourceMetadata;

export function stripExactSourceMetadata(text) {
    return tryDecodeExactSourceMetadata(text)?.visibleText ?? String(text);
}

export const strip_exact_source_metadata = stripExactSourceMetadata;

export function recoverEmbeddedExactSource(text) {
    return tryDecodeExactSourceMetadata(text)?.originalSource ?? null;
}

export const recover_embedded_exact_source = recoverEmbeddedExactSource;

export function hasEmbeddedExactSource(text) {
    return tryDecodeExactSourceMetadata(text) !== null;
}

export const has_embedded_exact_source = hasEmbeddedExactSource;
